improved test for shutdown wait limit

jjb opened the pull request:
before this change, if [the "killer thread" in Schedule#join_shutdown](https://github.com/jmettraux/rufus-scheduler/blob/master/lib/rufus/scheduler.rb#L555) was removed, no test would fail

jmettraux commented:
@jjb wrote:
> before this change, if the "killer thread" in Schedule#join_shutdown was removed, no test would fail

How so? There is one killer thread per worker thread. A killer thread is not created if the `shutdown` comes from a worker thread. In the test, the `shutdown` is ordered by the main thread, not a worker thread.

If I remember correctly, the "if a worker thread shuts down the scheduler, let trust it and let it terminate itself" comes from #304. I can try to remove the `work_thread == Thread.current ?` to see if the gh-304 are affected...

jmettraux commented:
I see no improvement in the pull request here.
The side pull request #316 says it's demonstrating a problem discussed here, but I see no problem and no problem discussion.

Of course, breaking the code will lead to problems, that is easy. Generally when one has a problem he demonstrates it by writing a script or a spec, or explaining the issue, not by breaking the code.

#314 also was starting in the wrong direction and I had to ask for a clear statement of the problem.

Please remember that my time is not infinite. If the time spent investigating an issue could be kept to a minimum, that'd leave me more time to actually fix the issue.

jjb commented:
Hi @jmettraux, thanks for all your attention to these issues and sorry that what I was doing here in this PR wasn't initially clear.

In #316, I removed the thread which kills worker jobs on a timer. This did _not_ result in any tests failing: https://travis-ci.org/github/jmettraux/rufus-scheduler/builds/752057812 (initially I had syntax errors in that branch, and the tests did fail, which you may have seen, so maybe that was confusing 😅 )

Here in #315, I write a test that will fail. To demonstrate this, I've added this test to #316 in an additional commit, and we see the test failures here: https://travis-ci.org/github/jmettraux/rufus-scheduler/builds/752059610

jmettraux commented:
Yes, I see that, removing the concurrent execution of the "join with time limit then kill" (max of (at most 2s, at most 2s)) ends up being not concurrent (at most 2s + at most 2s).

Anyway, I will add a test inspired by your test and then close.

jjb commented:
awesome, looks good!

<img width="755" alt="image" src="https://user-images.githubusercontent.com/6097/103325731-03177580-4a45-11eb-82b4-7e16949d3355.png">

jmettraux commented:
For the record.

> before this change, if the "killer thread" in Schedule#join_shutdown was removed, no test would fail

The extended test ensures that the timed joining and killing of the worker threads happens concurrently. The test as it was before @jjb suggestions was only testing against a single worker thread. It would not have failed, since one in concurrence lasts as long as one in sequence, while two in concurrence vs two in sequence...

Thanks a lot!

jjb commented:
great - thank you!
